src/runtime_src/core/common/drv/kds_core.c

```diff
@@ -1316,6 +1478,159 @@ int kds_del_context(struct kds_sched *kds, struct kds_client *client,
 		 pid_nr(client->pid), cu_domain, cu_idx);
 	return 0;
 }
+/* End Legacy Context for backward compartability */
+
+static int
+kds_initialize_cu_hw_ctx(struct kds_client *client, struct kds_client_cu_ctx *cu_ctx,
+		struct kds_client_cu_info *cu_info)
+{
+	if (!cu_ctx) {
+		kds_err(client, "No such Client HW Context available");
+		return -EINVAL;
+	}
+
+	cu_ctx->hw_ctx = cu_info->ctx;
+	cu_ctx->cu_domain = cu_info->cu_domain;
+	cu_ctx->cu_idx = cu_info->cu_idx;
+	cu_ctx->ref_cnt = 0;
+	cu_ctx->flags = cu_info->flags;
+
+	return 0;
+}
+
+struct kds_client_cu_ctx *
+kds_alloc_cu_hw_ctx(struct kds_client *client, struct kds_client_hw_ctx *hw_ctx,
+		struct kds_client_cu_info *cu_info)
+{
+	struct kds_client_cu_ctx *cu_ctx = NULL;
+
+	BUG_ON(!mutex_is_locked(&client->lock));
+
+	cu_ctx = kds_get_cu_hw_ctx(client, hw_ctx, cu_info);
+	if (IS_ERR(cu_ctx))
+		return NULL;
+
+	/* Valid CU context exists. Return this context here */
+	if (cu_ctx)
+		return cu_ctx;
```

Comment:
We returned an existing cu_ctx, but we did not bump up the cu_ctx->ref_cnt? What if, now, the caller try to free it? Or can caller know that it gets back an existing one, not a newly alloc'ed one, somehow?

Reply:
`
	/* Allocate a free CU context for the given CU index */
	cu_ctx = kds_alloc_cu_hw_ctx(client, hw_ctx, &cu_info);
	if (!cu_ctx) {
		userpf_err(xdev, "Allocation of CU context failed");
		mutex_unlock(&client->lock);
		return -EINVAL;
	}

        ret = kds_add_context(&XDEV(xdev)->kds, client, cu_ctx);
        if (ret) {
                kds_free_cu_ctx(client, cu_ctx);
		mutex_unlock(&client->lock);
		return ret;
	}`
Maybe you can help me understanding the code snippet above. When you call kds_alloc_cu_hw_ctx(), you'll get back a cu_ctx (assuming this is an existing one). Then you call kds_add_context(), which failed, ref count is not bumped up, then you call kds_free_cu_ctx(). Now kds_free_cu_ctx() will decrease the ref count (decrease without increase)?

Reply:
kds_free_cu_ctx() will not decrease the ref_cnt. 
 `abcdefgh`

Reply:
kds_free_cu_ctx will not decrease the ref_cnt. ref_cnt will be decreased only inside kds_del_context

int kds_free_cu_ctx(struct kds_client *client, struct kds_client_cu_ctx *cu_ctx)
{
        BUG_ON(!mutex_is_locked(&client->lock));

        if (!cu_ctx)
               return -EINVAL;

        if (cu_ctx->ref_cnt) {
                /* Reference count must be reset before free the context */
                kds_err(client, "Invalid CU Context requested to free");
                return -EINVAL;
        }

        list_del(&cu_ctx->link);
        vfree(cu_ctx);

        return 0;
}